src/classes/CON_CurrentGradeLevel_TDTM.cls

```diff
@@ -0,0 +1,136 @@
+            !TDTM_ProcessControl.getRecursionFlag(TDTM_ProcessControl.registeredTrigger.CON_CurrentGradeLevel_TDTM_After_Update)) {
+            if (listOld != NULL && listNew != NULL &&  listNew.size() > 0 &&  listOld.size() > 0) {
+                Map<Id, Contact> allIdsToOldContacts = new Map<Id, Contact>((List<Contact>)listOld);
+                Map<Id, Contact> allIdsToNewContacts = new Map<Id, Contact>((List<Contact>)listNew);
+                if (!TDTM_ProcessControl.getRecursionFlag(TDTM_ProcessControl.registeredTrigger.CON_CurrentGradeLevel_TDTM_createGradeEnrollments)) {
+                    createGradeEnrollments(allIdsToNewContacts, allIdsToOldContacts, dmlWrapper);
+                }    
+                updateGradeEnrollments(allIdsToNewContacts, allIdsToOldContacts, dmlWrapper);
+                // Set recursion flag to prevent recursion
+                TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.registeredTrigger.CON_CurrentGradeLevel_TDTM_After_Update, true);
+            }
+        }
+
+        // Set the recursion flags to false for next transaction
+        TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.registeredTrigger.CON_CurrentGradeLevel_TDTM_After_Insert, false);
+        TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.registeredTrigger.CON_CurrentGradeLevel_TDTM_After_Update, false);
+        TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.registeredTrigger.CON_CurrentGradeLevel_TDTM_createGradeEnrollments, false);
+
+        return dmlWrapper;
+    }
+    /*******************************************************************************************************
+    * @description This method creates Grade Enrollment records whenever a Contact with a Current Grade Level
+    * is inserted OR Contact's Current Grade Level is changed.
+    * @param allIdsToContacts A Map of Id to all new Contact Records
+    * @param allIdsToContactsOld A Map of Id to all old Contact Records
+    * @param dmlWrapper
+    ********************************************************************************************************/
+    private void createGradeEnrollments(Map<Id, Contact> allIdsToContacts, Map<Id, Contact> allIdsToContactsOld, dmlWrapper dmlWrapper) {
+        List<Grade_Enrollment__c> gradeEnrollmentsToInsert = new List<Grade_Enrollment__c>();
+        for (Contact eachContact : [SELECT Id, Current_Grade_Level__c FROM Contact WHERE (((Current_Grade_Level__c != NULL) AND (Current_Grade_Level__c != 'NotApplicable')
+                                                                                         AND (Current_Grade_Level__c != '--None--') AND (Current_Grade_Level__c != ''))
+                                                                                         AND Id IN :allIdsToContacts.keySet())]) {
+            if (eachContact != NULL && ((allIdsToContactsOld == NULL) || (allIdsToContactsOld != NULL && allIdsToContactsOld.get(eachContact.Id) != NULL
```

Comment:
We are doing alot of comparison here. Can we perhaps move this to a method, and and name the method to represent what this check is doing? Perhaps this is just checking that we need to update, that's fine, etc.

Reply:
Added a private method currentGradeEnrollmentChanged.